src/plugins/data/public/data_sources/datasource/types.ts

```diff
@@ -3,49 +3,79 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
+import { DataSource } from './datasource';
+
 /**
  * @experimental These interfaces are experimental and might change in future releases.
  */
 
-import { IndexPatternsService } from '../../index_patterns';
-import { DataSourceType } from '../datasource_services';
-
 export interface IndexPatternOption {
   title: string;
   id: string;
 }
 
-export interface IDataSourceMetaData {
+export interface IDataSourceGroup {
   name: string;
 }
 
-export interface IDataSourceGroup {
+export interface DataSetWithDataSource<T = []> {
+  ds: DataSource;
+  list: T;
+}
+
+export interface IDataSetParams<T = {}> {
+  query: T;
+}
+
+export interface IDataSourceQueryParams<T = {}> {
+  query: T;
+}
+
+export interface IDataSourceQueryResult<T = {}> {
+  data: T;
+}
+
+export interface DataSourceConnectionStatus {
+  status: string;
+  message: string;
+  error?: Error;
+}
+
+export interface IDataSourceSettings<T extends IDataSourceMetadata = IDataSourceMetadata> {
+  id: string;
+  type: string;
   name: string;
+  metadata: T;
 }
 
-export interface ISourceDataSet {
-  ds: DataSourceType;
-  data_sets: Array<string | IndexPatternOption>;
+export interface IDataSourceMetadata {
+  ui: IDataSourceUISettings;
 }
 
-// to-dos: add common interfaces for datasource
-// eslint-disable-next-line @typescript-eslint/no-empty-interface
-export interface IDataSetParams {}
+export interface IDataSourceUISelector {
+  displayDatasetsAsSource: boolean;
+}
 
-// eslint-disable-next-line @typescript-eslint/no-empty-interface
-export interface IDataSourceQueryParams {}
+export interface IDataSourceUISettings {
+  selector: IDataSourceUISelector;
+  label: string; // the display name of data source
+  groupType: DataSourceUIGroupType; // the group to which the data source belongs
+  typeLabel: string; // the display name of data source type
+  displayOrder?: number; // the order in which the data source should be displayed in selector
+  description?: string; // short description of your database
+  icon?: string; // uri of the icon
```

Comment:
i think tsdoc/jsdoc has `@member` or `/** xxx */` to document fields of interface, might be better so typescript language server would be aware of it

i'm not sure if `// xxx` works, if it does then ignore